zerver/actions/message_send.py

```diff
@@ -1460,6 +1461,21 @@ def check_private_message_policy(
         raise JsonableError(_("Direct messages are disabled in this organization."))
 
 
+def check_sender_can_access_recipients(
+    realm: Realm, sender: UserProfile, user_profiles: Sequence[UserProfile]
+) -> None:
+    if check_user_can_access_all_users(sender):
+        return
+
+    users_accessible_to_sender = set(get_accessible_user_ids(realm, sender))
```

Comment:
`get_accessible_user_ids` is quite expensive for this check; in the common case, `user_profiles` is going to be 1-2 users, and it should in theory be much cheaper to check facts like whether we've had any DMs with those users specifically than fetching the set of all users accessible to the sender.

Since sending DMs is a frequent operation, we probably do care about the latency this introduces for DMs sent by limited guests. It likely requires some messy plumbing, but I would expect a `user_profile__in=[small list]` filter on `subs_in_target_user_subscriptions_query` to be fairly significant too.

I think this is OK to fix in a follow-up; but I might add a TODO comment here before merging.

Reply:
Opened https://github.com/zulip/zulip/issues/27835 to track this.